File: src/main/java/com/polygloat/controllers/SourceController.java
package com.polygloat.controllers;

import com.polygloat.constants.Message;
import com.polygloat.dtos.request.EditSourceDTO;
import com.polygloat.dtos.request.SetTranslationsDTO;
import com.polygloat.dtos.response.SourceDTO;
import com.polygloat.exceptions.NotFoundException;
import com.polygloat.model.Permission;
import com.polygloat.model.Source;
import com.polygloat.service.SecurityService;
import com.polygloat.service.SourceService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import javax.validation.ValidationException;
import java.util.Set;

@RestController
@CrossOrigin(origins = "*")
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
@RequestMapping("/api/repository/{repositoryId}/sources")
public class SourceController implements IController {

    private final SourceService sourceService;
    private final SecurityService securityService;

    @PostMapping("/create")
    public void create(@PathVariable("repositoryId") Long repositoryId, @RequestBody @Valid SetTranslationsDTO dto) {
        Permission permission = securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.TRANSLATE);
        sourceService.createSource(permission.getRepository(), dto);
    }

    @PostMapping(value = "/edit")
    public void edit(@PathVariable("repositoryId") Long repositoryId, @RequestBody @Valid EditSourceDTO dto) {
        Permission permission = securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.EDIT);
        sourceService.editSource(permission.getRepository(), dto);
    }

    @GetMapping(value = "{id}")
    public SourceDTO get(@PathVariable("id") Long id) {
        Source source = sourceService.getSource(id).orElseThrow(NotFoundException::new);
        securityService.getAnyRepositoryPermission(source.getRepository().getId());
        return SourceDTO.builder().fullPathString(source.getName()).build();
    }

    @DeleteMapping(value = "/{id}")
    public void delete(@PathVariable Long id) {
        Source source = sourceService.getSource(id).orElseThrow(NotFoundException::new);
        securityService.checkRepositoryPermission(source.getRepository().getId(), Permission.RepositoryPermissionType.EDIT);
        sourceService.deleteSource(id);
    }

    @DeleteMapping(value = "")
    @Transactional
    public void delete(@PathVariable("repositoryId") Long repositoryId, @RequestBody Set<Long> ids) {
        securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.EDIT);
        for (Source source : sourceService.getSources(ids)) {
            if (!repositoryId.equals(source.getRepository().getId())) {
                throw new ValidationException(Message.SOURCE_NOT_FROM_REPOSITORY.getCode());
            }
            sourceService.deleteSources(ids);
        }
    }
}

File: src/main/java/io/polygloat/controllers/SourceController.kt
package io.polygloat.controllers

import io.polygloat.constants.Message
import io.polygloat.dtos.request.EditSourceDTO
import io.polygloat.dtos.request.SetTranslationsDTO
import io.polygloat.dtos.response.SourceDTO
import io.polygloat.exceptions.NotFoundException
import io.polygloat.model.Permission
import io.polygloat.service.KeyService
import io.polygloat.service.SecurityService
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.*
import javax.validation.Valid
import javax.validation.ValidationException

@RestController
@CrossOrigin(origins = ["*"])
@RequestMapping("/api/repository/{repositoryId}/sources")
open class SourceController(
        private val keyService: KeyService,
        private val securityService: SecurityService
) : IController {

    @PostMapping("/create")
    fun create(@PathVariable("repositoryId") repositoryId: Long?, @RequestBody dto: @Valid SetTranslationsDTO?) {
        val permission = securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.TRANSLATE)
        keyService.createSource(permission.repository!!, dto!!)
    }

    @PostMapping(value = ["/edit"])
    fun edit(@PathVariable("repositoryId") repositoryId: Long?, @RequestBody dto: @Valid EditSourceDTO?) {
        val permission = securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.EDIT)
        keyService.editSource(permission.repository!!, dto!!)
    }

    @GetMapping(value = ["{id}"])
    operator fun get(@PathVariable("id") id: Long?): SourceDTO {
        val key = keyService.getSource(id!!).orElseThrow { NotFoundException() }
        securityService.getAnyRepositoryPermission(key.repository!!.id)
        return SourceDTO(key.name)
    }

    @DeleteMapping(value = ["/{id}"])
    fun delete(@PathVariable id: Long?) {
        val key = keyService.getSource(id!!).orElseThrow { NotFoundException() }
        securityService.checkRepositoryPermission(key.repository!!.id, Permission.RepositoryPermissionType.EDIT)
        keyService.deleteSource(id)
    }

    @DeleteMapping(value = [""])
    @Transactional
    open fun delete(@PathVariable("repositoryId") repositoryId: Long, @RequestBody ids: Set<Long>?) {
        securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.EDIT)
        for (key in keyService.getSources(ids!!)) {
            if (repositoryId != key.repository!!.id) {
                throw ValidationException(Message.SOURCE_NOT_FROM_REPOSITORY.code)
            }
            keyService.deleteSources(ids)
        }
    }
}